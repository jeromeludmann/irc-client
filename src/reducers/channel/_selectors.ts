import { createSelector } from "reselect";
import { selectChannel } from "@app/reducers/server/_selectors";

export const selectMessages = createSelector(
  selectChannel,
  channel => channel.messages,
);

export const selectInput = createSelector(
  selectChannel,
  channel => channel.input,
);

export const selectValue = createSelector(selectInput, input => input.value);
