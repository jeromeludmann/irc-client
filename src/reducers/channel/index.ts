import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from "@app/reducers/channel/messages";
import {
  InputState,
  inputInitialState,
  reduceInput,
} from "@app/reducers/channel/input";
import {
  ActivityState,
  activityInitialState,
  reduceActivity,
} from "@app/reducers/channel/activity";
import { RoutedAction } from "@app/Route";
import { RouteState } from "@app/reducers/route";
import { UserState } from "@app/reducers/server/user";

export interface ChannelState {
  readonly messages: MessagesState;
  readonly input: InputState;
  readonly activity: ActivityState;
}

export const channelInitialState: ChannelState = {
  messages: messagesInitialState,
  input: inputInitialState,
  activity: activityInitialState,
};

export const reduceChannel = (
  channel = channelInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
): ChannelState => ({
  messages: reduceMessages(channel.messages, action, extraStates),
  input: reduceInput(channel.input, action),
  activity: reduceActivity(channel.activity, action, extraStates),
});
