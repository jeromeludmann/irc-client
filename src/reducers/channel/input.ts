import { Reducer, Action } from "redux";
import {
  UPDATE_INPUT_VALUE,
  UpdateInputValueAction,
  ENTER_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
  EnterInputValueAction,
} from "@app/actions/ui";

export interface InputState {
  value: string;
  dirtyValue: string;
  history: {
    values: string[];
    index: number;
  };
}

export const inputInitialState: InputState = {
  value: "",
  dirtyValue: "",
  history: { values: [], index: 0 },
};

const beginOfHistory = ({ history }: InputState) => history.index === 0;

const endOfHistory = ({ history }: InputState) =>
  history.index === history.values.length;

const handlers: {
  [action: string]: (input: InputState, action: Action) => InputState;
} = {
  [UPDATE_INPUT_VALUE]: (input, action: UpdateInputValueAction) => ({
    ...input,
    value: action.payload.value,
    dirtyValue: endOfHistory(input) ? action.payload.value : input.dirtyValue,
  }),

  [ENTER_INPUT_VALUE]: (input, action: EnterInputValueAction) => {
    const historyValues = [...input.history.values, action.payload.value];
    return {
      value: "",
      dirtyValue: "",
      history: { values: historyValues, index: historyValues.length },
    };
  },

  [GO_BACK_INPUT_HISTORY]: input =>
    beginOfHistory(input)
      ? input
      : {
          ...input,
          value: input.history.values[input.history.index - 1],
          history: { ...input.history, index: input.history.index - 1 },
        },

  [GO_FORWARD_INPUT_HISTORY]: input =>
    endOfHistory(input)
      ? input
      : {
          ...input,
          value:
            input.history.values[input.history.index + 1] || input.dirtyValue,
          history: { ...input.history, index: input.history.index + 1 },
        },
};

export const reduceInput: Reducer<InputState, Action> = (
  input = inputInitialState,
  action,
) =>
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](input, action)
    : input;
