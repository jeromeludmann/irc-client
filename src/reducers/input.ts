import { Action } from 'redux'
import {
  UPDATE_INPUT_VALUE,
  UpdateInputValueAction,
  ENTER_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
  EnterInputValueAction,
} from '@app/actions/ui'
import { CaseReducerMap } from '@app/utils/CaseReducerMap'

export type InputState = Readonly<{
  value: string
  dirtyValue: string
  history: Readonly<{
    values: string[]
    index: number
  }>
}>

type InputReducer = (input: InputState, action: Action) => InputState

export const inputInitialState: InputState = {
  value: '',
  dirtyValue: '',
  history: { values: [], index: 0 },
}

const beginOfHistory = (input: InputState) => {
  return input.history.index === 0
}

const endOfHistory = (input: InputState) => {
  return input.history.index === input.history.values.length
}

const caseReducers: CaseReducerMap<InputReducer> = {
  [ENTER_INPUT_VALUE]: (input, action: EnterInputValueAction) => {
    const historyValues = [...input.history.values, action.payload.value]
    return {
      value: '',
      dirtyValue: '',
      history: { values: historyValues, index: historyValues.length },
    }
  },

  [GO_BACK_INPUT_HISTORY]: input =>
    beginOfHistory(input)
      ? input
      : {
          ...input,
          value: input.history.values[input.history.index - 1],
          history: { ...input.history, index: input.history.index - 1 },
        },

  [GO_FORWARD_INPUT_HISTORY]: input =>
    endOfHistory(input)
      ? input
      : {
          ...input,
          value:
            input.history.values[input.history.index + 1] || input.dirtyValue,
          history: { ...input.history, index: input.history.index + 1 },
        },

  [UPDATE_INPUT_VALUE]: (input, action: UpdateInputValueAction) => ({
    ...input,
    value: action.payload.value,
    dirtyValue: endOfHistory(input) ? action.payload.value : input.dirtyValue,
  }),
}

export const reduceInput: InputReducer = (input, action) =>
  action.type in caseReducers ? caseReducers[action.type](input, action) : input
