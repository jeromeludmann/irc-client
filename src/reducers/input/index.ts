import {
  HistoryState,
  historyInitialState,
  reduceHistory,
} from "@app/reducers/input/history";
import {
  inputValueInitialState,
  reduceInputValue,
} from "@app/reducers/input/value";
import {
  inputDirtyValueInitialState,
  reduceInputDirtyValue,
} from "@app/reducers/input/dirtyValue";
import { Reducer, Action } from "redux";

export interface InputState {
  readonly value: string;
  readonly dirtyValue: string;
  readonly history: HistoryState;
}

export const inputInitialState: InputState = {
  value: inputValueInitialState,
  dirtyValue: inputDirtyValueInitialState,
  history: historyInitialState,
};

export const reduceInput: Reducer<InputState, Action> = (
  input = inputInitialState,
  action,
) => ({
  value: reduceInputValue(input.value, action, { input }),
  dirtyValue: reduceInputDirtyValue(input.dirtyValue, action, { input }),
  history: reduceHistory(input.history, action),
});
