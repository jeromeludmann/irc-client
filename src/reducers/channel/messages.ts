import { Action } from "redux";
import { mapReducer } from "@app/reducers/_map";
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  ConnectionClosedAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
} from "@app/actions/socket";
import {
  ErrorAction,
  JoinAction,
  PartAction,
  PrivmsgAction,
  NoticeFromServerAction,
  NoticeFromChannelAction,
  NoticeFromUserAction,
  PingFromServerAction,
  ERROR,
  JOIN,
  NOTICE_FROM_SERVER,
  NOTICE_FROM_CHANNEL,
  NOTICE_FROM_USER,
  PART,
  PRIVMSG,
  PING_FROM_SERVER,
} from "@app/actions/messages";

export type MessagesState = string[];

export const messagesInitialState: MessagesState = [];

type MessagesReducer<A = Action> = (
  messages: MessagesState,
  action: A,
) => MessagesState;

const connectionFailed: MessagesReducer<ConnectionFailedAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.message];
};

const connectionClosed: MessagesReducer<ConnectionClosedAction> = messages => {
  return [...messages, "Disconnected from remote host."];
};

const error: MessagesReducer<ErrorAction> = (messages, action) => {
  return [...messages, action.payload.message];
};

const join: MessagesReducer<JoinAction> = (messages, action) => {
  const { user, channel } = action.payload;
  const msg = `${user.nick} has joined ${channel}`;
  return [...messages, msg];
};

const part: MessagesReducer<PartAction> = (messages, action) => {
  const { user, channel, message } = action.payload;
  const baseMsg = `${user.nick} has left ${channel}`;
  const msg = message ? `${baseMsg} (${message})` : baseMsg;
  return [...messages, msg];
};

const privmsg: MessagesReducer<PrivmsgAction> = (messages, action) => {
  const { user, text } = action.payload;
  const msg = `${user.nick}: ${text}`;
  return [...messages, msg];
};

const raw: MessagesReducer<RawMessagesAction> = (messages, action) => {
  return [...messages, ...action.payload.messages];
};

const noticeFromServer: MessagesReducer<NoticeFromServerAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.text];
};

const noticeFromChannel: MessagesReducer<NoticeFromChannelAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  const { channelKey } = action.route;
  return [...messages, `-${user.nick}/${channelKey}- ${text}`];
};

const noticeFromUser: MessagesReducer<NoticeFromUserAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  return [...messages, `-${user.nick}- ${text}`];
};

const serverPing: MessagesReducer<PingFromServerAction> = messages => {
  return [...messages, "Ping?"];
};

const map: { [action: string]: MessagesReducer } = {
  [CONNECTION_FAILED]: connectionFailed,
  [CONNECTION_CLOSED]: connectionClosed,
  [ERROR]: error,
  [JOIN]: join,
  [NOTICE_FROM_SERVER]: noticeFromServer,
  [NOTICE_FROM_CHANNEL]: noticeFromChannel,
  [NOTICE_FROM_USER]: noticeFromUser,
  [PART]: part,
  [PRIVMSG]: privmsg,
  [RAW_MESSAGES_RECEIVED]: raw,
  [PING_FROM_SERVER]: serverPing,
};

export const reduceMessages = mapReducer<MessagesState>(map);
