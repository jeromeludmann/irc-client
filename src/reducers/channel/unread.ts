import { RouteState } from "@app/reducers/route";
import { SWITCH_WINDOW, SwitchWindowAction } from "@app/actions/ui";
import { RAW_MESSAGES_RECEIVED } from "@app/actions/socket";
import { UserState } from "@app/reducers/server/user";
import { mapReducer } from "@app/reducers/_map";
import { Action } from "redux";
import { RoutedAction, BROADCAST_ACTIVE } from "@app/Route";
import {
  JoinAction,
  JOIN,
  PART,
  PRIVMSG,
  NOTICE_FROM_SERVER,
  NOTICE_FROM_CHANNEL,
  NOTICE_FROM_USER,
} from "@app/actions/messages";

export type UnreadState = boolean;

export const unreadInitialState: UnreadState = false;

type UnreadReducer<A = Action> = (
  state: UnreadState,
  action: A,
  extraStates: { route: RouteState; user: UserState },
) => UnreadState;

const join: UnreadReducer<JoinAction> = (_, action, extraStates) =>
  action.payload.user.nick !== extraStates.user.nick;

const switchRoute: UnreadReducer<SwitchWindowAction> = () => false;

const anyActivity: UnreadReducer<RoutedAction> = (_, action, extraStates) => {
  const sameRoute =
    action.route.serverKey === extraStates.route.serverKey &&
    action.route.channelKey === extraStates.route.channelKey;
  return !sameRoute && action.route.channelKey !== BROADCAST_ACTIVE;
};

const map: { [action: string]: UnreadReducer } = {
  [JOIN]: join,
  [SWITCH_WINDOW]: switchRoute,
  [PART]: anyActivity,
  [PRIVMSG]: anyActivity,
  [NOTICE_FROM_SERVER]: anyActivity,
  [NOTICE_FROM_CHANNEL]: anyActivity,
  [NOTICE_FROM_USER]: anyActivity,
  [RAW_MESSAGES_RECEIVED]: anyActivity,
};

export const reduceUnread = mapReducer<
  UnreadState,
  { route: RouteState; user: UserState }
>(map);
