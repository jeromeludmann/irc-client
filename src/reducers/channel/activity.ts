import { RoutedAction, BROADCAST_ACTIVE } from "@app/Route";
import { RouteState } from "@app/reducers/route";
import { SWITCH_WINDOW } from "@app/actions/ui";
import { UserState } from "@app/reducers/server/user";
import { JoinAction, JOIN } from "@app/actions/messages";

export type ActivityState = boolean;

export const activityInitialState: ActivityState = false;

type ActivityReducer<A = RoutedAction> = (
  state: ActivityState,
  action: A,
  extraStates: { route: RouteState; user: UserState },
) => ActivityState;

const reducers: { [action: string]: ActivityReducer } = {
  [JOIN]: (_, action: JoinAction, extraStates) =>
    action.payload.user.nick !== extraStates.user.nick,

  [SWITCH_WINDOW]: () => false,
};

const hasAnyActivity: ActivityReducer<RoutedAction> = (
  _,
  action,
  extraStates,
) => {
  const sameRoute =
    action.route.serverKey === extraStates.route.serverKey &&
    action.route.channelKey === extraStates.route.channelKey;
  return !sameRoute && action.route.channelKey !== BROADCAST_ACTIVE;
};

export const reduceActivity = (
  activity: ActivityState = activityInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState; user: UserState },
) =>
  reducers.hasOwnProperty(action.type)
    ? reducers[action.type](activity, action, extraStates)
    : hasAnyActivity(activity, action, extraStates);
