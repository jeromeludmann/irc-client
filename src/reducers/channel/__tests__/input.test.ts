import { reduceInput, inputInitialState } from "@app/reducers/channel/input";
import {
  updateInputValue,
  enterInputValue,
  goBackInputHistory,
  goForwardInputHistory,
} from "@app/actions/ui";

describe("reduce input", () => {
  const historyValues = ["hello", "world"];

  describe("update input value", () => {
    it("anywhere in history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 1, values: historyValues },
          },
          updateInputValue("hello"),
        ),
      ).toMatchSnapshot();
    });

    it("at the end of history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 2, values: historyValues },
          },
          updateInputValue("hello"),
        ),
      ).toMatchSnapshot();
    });
  });

  describe("enter input value", () => {
    expect(reduceInput(undefined, enterInputValue("hello"))).toMatchSnapshot();
  });

  describe("go back input history", () => {
    it("at the begin of history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 0, values: historyValues },
          },
          goBackInputHistory(),
        ),
      ).toMatchSnapshot();
    });

    it("anywhere in history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 1, values: historyValues },
          },
          goBackInputHistory(),
        ),
      ).toMatchSnapshot();
    });
  });

  describe("go forward input history", () => {
    it("anywhere in the history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 1, values: historyValues },
          },
          goForwardInputHistory(),
        ),
      ).toMatchSnapshot();
    });

    it("at the end of history", () => {
      expect(
        reduceInput(
          {
            ...inputInitialState,
            history: { index: 2, values: historyValues },
          },
          goForwardInputHistory(),
        ),
      ).toMatchSnapshot();
    });
  });
});
