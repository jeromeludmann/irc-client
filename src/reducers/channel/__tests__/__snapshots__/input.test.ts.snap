// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 1,
    "values": Array [
      "hello",
    ],
  },
  "value": "",
}
`;

exports[`reduce input go back input history anywhere in history 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 0,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "hello",
}
`;

exports[`reduce input go back input history at the begin of history 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 0,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "",
}
`;

exports[`reduce input go forward input history anywhere in the history 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 2,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "",
}
`;

exports[`reduce input go forward input history at the end of history 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 2,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "",
}
`;

exports[`reduce input update input value anywhere in history 1`] = `
Object {
  "dirtyValue": "",
  "history": Object {
    "index": 1,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "hello",
}
`;

exports[`reduce input update input value at the end of history 1`] = `
Object {
  "dirtyValue": "hello",
  "history": Object {
    "index": 2,
    "values": Array [
      "hello",
      "world",
    ],
  },
  "value": "hello",
}
`;
